docs/GDPs/gdp-0003-duck-storage.rst

```diff
@@ -0,0 +1,750 @@
+nevertheless work, with a well-defined default behavior.
+
+Finally, without internally keeping information about the semantic meaning of dimensions, e.g. the
+best layout and proper broadcasting for the resulting storage can not be determined. Further,
+implementations would depend on the availability of a library implementing the operations for a
+given device. We have already observed performance problems when using cupy on AMD hardware. For
+future hardware, these libraries might be entirely unavailable. Therefore, we will not commit to
+supporting such operations.
+
+Backward Compatibility
+----------------------
+
+The implementation of this GDP breaks the integration with external libraries which require a NumPy
+`ndarray` subclass. Further, we propose some API changes like renaming or repurposing of keyword
+arguments and attributes. These changes are expected to break most of existing user codes. However,
+existing GT4Py functionality will remain or be extended and thus updating codebases to the new
+interface amounts mostly to updating attribute and keyword argument names.
+
+The removal of universal functions will require codes relying on ufuncs to be re-written by
+implementing these using third-party libraries.
+
+
+Detailed Description
+--------------------
+
+Functionality
+^^^^^^^^^^^^^
+
+We chose to use internally NumPy and CuPy `ndarrays` to store CPU and GPU buffers, respectively,
+since they are standard and reliable components of the Python ecosystem. We rely on those libraries
+to implement part of the functionality, like :code:`__setitem__` and :code:`__getitem__`, by
+forwarding the calls to the appropriate NumPy/CuPy function call. As a consequence, support for
+`dtypes` and other functionality is restricted to the common denominator of CuPy and NumPy.
+
+.. note:: In this document, references to NumPy and CuPy objects or functions use the standard
+    shorted prefiex for these libraries, that is :code:`np.` for NumPy and :code:`cp.` for CuPy.
+
+Data Interface
+^^^^^^^^^^^^^^
+
+Objects implementing the data interface have the attribute :code:`__gt_data_interface__`. This
+attribute is a dictionary which maps `device identifiers` to a dictionary similar to those
+defined as the :code:`__array_interface__` and the :code:`__cuda_array_interface__`.
+
+A device identifier can be one of:
+
++ :code:`None` denoting a buffer in main memory
++ :code:`"gpu"` denoting a buffer on a GPU.
+
+The mapped dictionary in turn must contain the following keys and mapped objects of pairs. Their
+meaning is the same as in the NumPy :code:`__array_interface__` and the
+:code:`__cuda_array_interface__`:
+
++ :code:`"shape": Tuple[int, ...]`
++ :code:`"typestr": str`
++ :code:`"data": Tuple[int, bool]`
++ :code:`"strides": Tuple[int, ...]`
+
+In Addition, the following optional keys can be contained:
+
++ :code:`"acquire": Optional[Callable[[], Any]]` Is called on all objects that are passed to a
```

Comment:
Including the origin and domain of the stencil call as arguments is valuable information for this method and could be used for a `DistributedStorage` implementation. Additionally for asynchronous exchanges a `release` method would be required (`acquire` waits if dirty, `release` initiates).

@mcgibbon Since you were expecting arguments, which ones would that be?

Reply:
The doc says "acquire is called on all objects that are passed to a stencil", which to me says that the objects are passed in as arguments to acquire. Is that correct? It seems that either the type hint should indicate it takes in an argument (an object passed in to a stencil), or the line after should be updated.

Reply:
I see. I read this as "acquire is called _for_ all objects that are passed to a stencil" which is in line with the signature. Maybe @gronerl can clarify, what is actually meant. Passing all fields, could actually be useful for delaying halo exchanges if the order of the `acquire` calls is clear.